lib/flow_runner/spec/block.ex

```diff
@@ -222,62 +215,56 @@ defmodule FlowRunner.Spec.Block do
     with {:ok, user_input} <-
            block_module.evaluate_outgoing(container, flow, block, context, user_input),
          # Process any user input we have been given.
-         {:ok, context} <-
-           Block.evaluate_user_input(
-             block,
-             context,
-             user_input
-           ),
-         {:ok, context, block} <-
-           Block.fetch_next_block(block, flow, context) do
+         {:ok, context} <- evaluate_user_input(block, context, user_input),
+         {:ok, context, block} <- fetch_next_block(block, flow, context) do
       {:ok, context, block}
     else
       {:invalid, reason} ->
         Logger.info("Fetching default block because #{reason}")
-        {:ok, context} = Block.evaluate_user_input(block, context, user_input)
-        Block.fetch_default_block(block, flow, context)
 
-      err ->
-        err
+        {:ok, context} = evaluate_user_input(block, context, user_input)
+
+        fetch_default_block(block, flow, context)
     end
   end
 
+  @spec fetch_default_block(Block.t(), Flow.t(), Context.t()) ::
+          {:error, iodata} | {:ok, Context.t(), Block.t() | nil}
   def fetch_default_block(block, %Flow{} = flow, %Context{} = context) do
-    {:ok, %Exit{destination_block: destination_block}} =
-      Block.evaluate_default_exit(block, context)
-
-    if destination_block == "" || destination_block == nil do
-      {:ok, %Context{context | finished: true}, nil}
-    else
-      case Flow.fetch_block(flow, destination_block) do
-        {:ok, next_block} -> {:ok, context, next_block}
-        {:error, reason} -> {:error, reason}
-      end
+    case evaluate_default_exit(block) do
+      {:ok, %Exit{destination_block: destination_block}}
+      when is_nil(destination_block) or destination_block == "" ->
+        {:ok, %Context{context | finished: true}, nil}
+
+      {:ok, %Exit{destination_block: destination_block}} ->
+        with {:ok, next_block} <- Flow.fetch_block(flow, destination_block) do
+          {:ok, context, next_block}
+        end
+
+      {:error, reason} ->
+        {:error, reason}
     end
```

Comment:
This is the actual fix. The rest of the changes are code refactors to make it simpler.